runtime/src/iree/hal/drivers/hip/hip_device_group_device.c

```diff
@@ -0,0 +1,1535 @@
+    iree_hal_stream_tracing_device_interface_t* base_device_interface,
+    iree_hal_stream_tracing_native_event_t base_event) {
+  iree_hal_hip_tracing_device_group_device_interface_t* device_interface =
+      (iree_hal_hip_tracing_device_group_device_interface_t*)
+          base_device_interface;
+
+  return IREE_HIP_RESULT_TO_STATUS(
+      device_interface->hip_symbols,
+      hipEventRecord(
+          (hipEvent_t)base_event,
+          (hipStream_t)device_interface->device_context->hip_dispatch_stream));
+}
+
+iree_status_t
+iree_hal_hip_tracing_device_group_device_interface_add_graph_event_record_node(
+    iree_hal_stream_tracing_device_interface_t* base_device_interface,
+    iree_hal_stream_tracing_native_graph_node_t* out_node,
+    iree_hal_stream_tracing_native_graph_t graph,
+    iree_hal_stream_tracing_native_graph_node_t* dependency_nodes,
+    size_t dependency_nodes_count,
+    iree_hal_stream_tracing_native_event_t event) {
+  iree_hal_hip_tracing_device_group_device_interface_t* device_interface =
+      (iree_hal_hip_tracing_device_group_device_interface_t*)
+          base_device_interface;
+
+  return IREE_HIP_RESULT_TO_STATUS(
+      device_interface->hip_symbols,
+      hipGraphAddEventRecordNode((hipGraphNode_t*)out_node, (hipGraph_t)graph,
+                                 (hipGraphNode_t*)dependency_nodes,
+                                 dependency_nodes_count, (hipEvent_t)event));
+}
+
+static iree_hal_hip_device_group_device_t*
+iree_hal_hip_device_group_device_cast(iree_hal_device_t* base_value) {
+  IREE_HAL_ASSERT_TYPE(base_value, &iree_hal_hip_device_group_device_vtable);
+  return (iree_hal_hip_device_group_device_t*)base_value;
+}
+
+static iree_hal_hip_device_group_device_t*
+iree_hal_hip_device_group_device_cast_unsafe(iree_hal_device_t* base_value) {
+  return (iree_hal_hip_device_group_device_t*)base_value;
+}
+
+IREE_API_EXPORT void iree_hal_hip_device_group_device_params_initialize(
+    iree_hal_hip_device_params_t* out_params) {
+  memset(out_params, 0, sizeof(*out_params));
+  out_params->arena_block_size = 32 * 1024;
+  out_params->event_pool_capacity = 32;
+  out_params->queue_count = 1;
+  out_params->command_buffer_mode = IREE_HAL_HIP_COMMAND_BUFFER_MODE_STREAM;
+  out_params->stream_tracing = 0;
+  out_params->async_allocations = true;
+  out_params->allow_inline_execution = false;
+}
+
+static iree_status_t iree_hal_hip_device_group_device_check_params(
+    const iree_hal_hip_device_params_t* params) {
+  if (params->arena_block_size < 4096) {
+    return iree_make_status(IREE_STATUS_INVALID_ARGUMENT,
+                            "arena block size too small (< 4096 bytes)");
+  }
+  if (params->queue_count == 0) {
+    return iree_make_status(IREE_STATUS_INVALID_ARGUMENT,
+                            "at least one queue is required");
+  }
+  return iree_ok_status();
+}
+
+static iree_status_t iree_hal_hip_device_group_device_initialize_internal(
```

Comment:
Tracing should be added to all code unless by exception for performance or noise issues - especially init/deinit or anything that may do an alloc should be traced.